src/table/use-sticky-columns.tsx

```diff
@@ -0,0 +1,230 @@
+
+  const updateCellOffsets = useCallback(() => {
+    // Calculate widths of all previous siblings of each table cell in the `tableCellRefs` array
+    let firstColumnsOffsets = tableCellRefs
+      .map(ref => (ref?.current?.previousSibling as HTMLTableCellElement)?.offsetWidth)
+      .filter(Boolean);
+    // Calculate cumulative widths of previous siblings to get the total offset of each of the cells
+    firstColumnsOffsets = firstColumnsOffsets.map((elem, index) =>
+      firstColumnsOffsets.slice(0, index + 1).reduce((a, b) => a + b)
+    );
+
+    // Calculate widths of all next siblings of each table cell in the `tableCellRefs` array
+    let lastColumnsOffsets = tableCellRefs.map(ref => (ref?.current?.nextSibling as HTMLTableCellElement)?.offsetWidth);
+    lastColumnsOffsets = lastColumnsOffsets.filter(Boolean).reverse();
+    // Calculate cumulative widths of next siblings to get the the total offset of each of the cells
+    lastColumnsOffsets = lastColumnsOffsets
+      .map((elem, index) => lastColumnsOffsets.slice(0, index + 1).reduce((a, b) => a + b))
+      .reverse();
+
+    setCellOffsets({ first: [0, ...firstColumnsOffsets], last: [...lastColumnsOffsets, 0] });
+  }, [tableCellRefs]);
+
+  useLayoutEffect(() => {
+    // Request animation frame to make sure effect runs after the browser's automatic table layout algorithm
+    const frameId = requestAnimationFrame(() => {
+      updateCellOffsets();
+    });
+
+    return () => {
+      cancelAnimationFrame(frameId);
+    };
+  }, [updateCellOffsets]);
+
+  return { cellOffsets, updateCellOffsets };
+};
+
+interface GetStickyClassNamesProps {
+  styles: { [key: string]: string };
+  isSticky: boolean;
+  isLastStickyLeft: boolean;
+  isLastStickyRight: boolean;
+  isStuckToTheLeft: boolean;
+  isStuckToTheRight: boolean;
+}
+
+export function getStickyClassNames({
+  styles,
+  isSticky,
+  isLastStickyLeft,
+  isLastStickyRight,
+  isStuckToTheLeft,
+  isStuckToTheRight,
+}: GetStickyClassNamesProps) {
+  return clsx({
+    [styles['sticky-cell']]: isSticky,
+    [styles['sticky-cell-last-left']]: isStuckToTheLeft && isLastStickyLeft,
+    [styles['sticky-cell-last-right']]: isStuckToTheRight && isLastStickyRight,
+  });
+}
+
+export const useStickyColumns = ({
+  containerWidth,
+  hasSelection,
+  isWrapperScrollable,
+  stickyColumns,
+  tableRefObject,
+  visibleColumnsLength,
+}: StickyColumnParams) => {
+  // Check if there are any sticky columns
+  const noStickyColumns = !stickyColumns?.first && !stickyColumns?.last;
+  const columnsLengthWithSelection = visibleColumnsLength + (hasSelection ? 1 : 0);
+
+  const [shouldDisable, setShouldDisable] = useState<boolean>(noStickyColumns);
+  const [tableCellRefs, setTableCellRefs] = useState<Array<React.RefObject<HTMLTableCellElement>>>([]);
+  const { cellOffsets, updateCellOffsets } = useCellOffsets(tableCellRefs);
+
+  // Compute left table padding
+  const table = tableRefObject.current;
```

Comment:
Although it is very common in the table's code, I think passing ref objects is an anti-pattern. These objects are mutable and it is very difficult to track their lifecycle - when it is updated and when it is read.

Instead, the hook can accept e.g. `tablePadding: { left: number, right: number }`.

Also, in the first render the table is going to be null and unless other props like `containerWidth` of `isWrapperScrollable` would change, the state that relates on the paddings won't update. I think the paddings require their own state, like:

```
const [tablePadding, setTablePadding] = useState({ left: 0, right: 0 })

useEffect(() => {
  if (!tableRefObject.current) {
    return;
  }
  const left = parseInt(getComputedStyle(tableRefObject.current).paddingLeft) || 0;
  const right = parseInt(getComputedStyle(tableRefObject.current).paddingRight) || 0;
  setTablePadding({ left, right })
}, [])
```